### ISSUE TITLE
cannot open old wallet file, convert_version_24 fails

### ISSUE THREAD

**Issue by tippet** [spesmilo/electrum#6863]
Thanks for any help.  
started in terminal:  ~/.local/bin $ ./electrum
```
qt5ct: using qt5ct plugin
qt5ct: D-Bus system tray: yes
E | daemon.Daemon | GUI raised exception: ValueError('too many values to unpack (expected 2)'). shutting down.
E | __main__ | daemon.run_gui errored
Traceback (most recent call last):
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/gui/qt/__init__.py", line 245, in wrapper
    return func(self, *args, **kwargs)
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/gui/qt/__init__.py", line 269, in start_new_window
    wallet = self._start_wizard_to_select_or_create_wallet(path)
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/gui/qt/__init__.py", line 314, in _start_wizard_to_select_or_create_wallet
    db = WalletDB(storage.read(), manual_upgrades=False)
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/wallet_db.py", line 72, in __init__
    self.load_data(raw)
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/wallet_db.py", line 105, in load_data
    self.upgrade()
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/util.py", line 408, in <lambda>
    return lambda *args, **kw_args: do_profile(args, kw_args)
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/util.py", line 404, in do_profile
    o = func(*args, **kw_args)
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/wallet_db.py", line 174, in upgrade
    self._convert_version_24()
  File "/home/tippet/.local/lib/python3.8/site-packages/electrum/wallet_db.py", line 531, in _convert_version_24
    for ser, v in l:
ValueError: too many values to unpack (expected 2)
 
Additional information
Electrum version: 4.0.8 
Python version: 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0] 
Operating system: Linux-5.4.0-58-generic-x86_64-with-glibc2.29 
Wallet type: standard 
Locale: en_US 
```

**Comment by SomberNight** [comment]
probably duplicate of https://github.com/spesmilo/electrum/issues/6349

do you have any information about using approx what version or when the wallet was created, and approx which versions of Electrum had opened the wallet?

It would also be good to know what 'seed_version' is in the wallet file -- i.e. what storage version are you trying to upgrade from.

**Comment by ecdsa** [comment]
I think this is because `convert_version_24` is performing inplace modifications of txi and txo.
we should probably do this:

```diff --git a/electrum/wallet_db.py b/electrum/wallet_db.py
index 49bc11ab1..f60ec2263 100644
--- a/electrum/wallet_db.py
+++ b/electrum/wallet_db.py
@@ -524,7 +524,7 @@ class WalletDB(JsonDB):
         self.data['channels'] = { x['channel_id']: x for x in channels }
         # convert txi & txo
         txi = self.get('txi', {})
-        for tx_hash, d in txi.items():
+        for tx_hash, d in list(txi.items()):
             d2 = {}
             for addr, l in d.items():
                 d2[addr] = {}
@@ -533,7 +533,7 @@ class WalletDB(JsonDB):
             txi[tx_hash] = d2
         self.data['txi'] = txi
         txo = self.get('txo', {})
-        for tx_hash, d in txo.items():
+        for tx_hash, d in list(txo.items()):
             d2 = {}
             for addr, l in d.items():
                 d2[addr] = {}
```

**Comment by SomberNight** [comment]
@ecdsa that sounds plausible however I cannot reproduce :/
I have even tested with a quite large wallet (txi dict around 5 MB), both on python 3.6 and 3.8.
Also, [this stackoverflow](https://stackoverflow.com/q/2315520) post says making a copy is not needed.

I guess we can commit that change and see what happens though.

**Comment by tippet** [comment]
Thanks for answering so quickly.  
That said, I backed up wallet_db.py, made the changes indicated, and got the same error response when opening an older wallet.
Opened my old wallets, and see the seed versions is listed as 18.  
The wallet which failed was originally from 2016 (lucky me), probably created with version 2.7.11 - as there are indications I once had that version.  More recently, version 3.38 was in use.

**Comment by tippet** [comment]
Updated the 'closed' issue with below:

Thanks for answering so quickly.
That said, I backed up wallet_db.py, made the changes indicated, and got
the same error response when opening an older wallet.
Opened my old wallets, and see the seed versions are listed as 18.
The wallet which failed was originally from 2016 (lucky me), probably
created with version 2.7.11 - as there are indications I once had that
version.  More recently, version 3.38 was in use.


On 12/18/20 3:22 AM, ghost43 wrote:

probably duplicate of #6349
<https://github.com/spesmilo/electrum/issues/6349>

do you have any information about using approx what version or when the
wallet was created, and approx which versions of Electrum had opened the
wallet?

It would also be good to know what 'seed_version' is in the wallet file --
i.e. what storage version are you trying to upgrade from.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/spesmilo/electrum/issues/6863#issuecomment-747967259>,
or unsubscribe

**Comment by SomberNight** [comment]
Make sure you have backups of the wallet file itself.

Can it still be opened with Electrum 3.3.8?

How large is the wallet file on disk?

**Comment by tippet** [comment]
neither old versions work as is:

~/_Electrum-3.3.8 $ ./run_electrum 
/home/tippet/_Electrum-3.3.8/packages/aiohttp/helpers.py:105: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
  def noop(*args, **kwargs):  # type: ignore
qt5ct: using qt5ct plugin
qt5ct: D-Bus system tray: yes

~/_Electrum-2.7.11 $ ./electrum
Traceback (most recent call last):
  File "/home/tippet/_Electrum-2.7.11/lib/plugins.py", line 80, in load_plugins
    self.load_plugin(name)
  File "/home/tippet/_Electrum-2.7.11/lib/plugins.py", line 99, in load_plugin
    p = loader.load_module(full_name)
  File "/usr/lib/python2.7/pkgutil.py", line 243, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/home/tippet/_Electrum-2.7.11/plugins/exchange_rate/qt.py", line 2, in <module>
    from PyQt4.QtGui import *
ImportError: No module named PyQt4.QtGui
Error: Could not import PyQt4 on Linux systems, you may try 'sudo apt-get install python-qt4'

file size of wallet 62K

**Comment by tippet** [comment]
My wallet files must have been corrupted.  4.0.8 did open an older backup copy successfully.

**Comment by scott-hand** [comment]
I'm not going to do a PR because I don't really know enough about this project to know if this is a general solution or just a patch for me. But this change fixed it and I was able to open my 2.4 version wallet:

```
index b37a8b035..1de326a78 100644
--- a/electrum/wallet_db.py
+++ b/electrum/wallet_db.py
@@ -528,7 +528,7 @@ class WalletDB(JsonDB):
             d2 = {}
             for addr, l in d.items():
                 d2[addr] = {}
-                for ser, v in l:
+                for ser, v in l.items():
                     d2[addr][ser] = v
             txi[tx_hash] = d2
         self.data['txi'] = txi
@@ -537,7 +537,7 @@ class WalletDB(JsonDB):
             d2 = {}
             for addr, l in d.items():
                 d2[addr] = {}
-                for n, v, cb in l:
+                for n, (v, cb) in l.items():
                     d2[addr][str(n)] = (v, cb)
             txo[tx_hash] = d2
         self.data['txo'] = txo
```

Hope this can help someone else in a similar spot.
